### PR TITLE
More Changes to #4287 - Help

### PR DESCRIPTION
**Pull request by TheWitness**
This change does a few things:

- Make a new setting to force either local or use docs.cacti.net
- Move response to JSON so we can create messages in HTML
- More selective messages depending on the error
- Updates language files again

### CODE REVIEW

**Comment by TheWitness** [comment]
Example not found on Web Site message:

image

**Comment by TheWitness** [comment]
Unable to reach site message:

image

**Comment by TheWitness** [comment]
New setting

image

**Comment by TheWitness** [comment]
Local install and file missing.

image

File: help.php [hunk 1]
```diff
@@ -29,14 +29,76 @@
...
+		$options = array(
+			'http' => array (
+				'method'            => 'GET'
+			),
+			'ssl' => array(
+				'verify_peer'       => false,
+				'verify_peer_name'  => false,
+				'allow_self_signed' => true,
+				'timeout'           => 2,
+			)
```
Comment (netniV) [inline help.php:44]: This should be used for local servers, not for our server.  If our server cert is not valid, something dodgy is going on.

File: help.php [hunk 2]
```diff
@@ -29,14 +29,76 @@
...
+		);
+
+		$context  = stream_context_create($options);
+		$url      = 'https://docs.cacti.net/' . $page;
+		$response = get_headers($url, true, $context);
+
+		if ($response !== false && $response[0] == 'HTTP/1.1 200 OK') {
+			$contents = file_get_contents($url, false, $context);
+
+			if ($contents != '' && !preg_match('/does not appear to exist/i', $contents)) {
```
Comment (netniV) [inline help.php:54]: When this is returned, do we not also st 404 as the status code?  Using a hard-coded string could cause help to fail between versions if that changes or a different language is usd.

File: help.php [hunk 3]
```diff
@@ -29,14 +29,76 @@
...
+				print json_encode(
+					array(
+						'status'   => 'Success',
+						'location' => 'https://docs.cacti.net/' . $page
+					)
+				);
+			} elseif ($contents != '' && preg_match('/does not appear to exist/i', $contents)) {
```
Comment (netniV) [inline help.php:61]: Hard coding, see above.

File: help.php [hunk 4]
```diff
@@ -29,14 +29,76 @@
...
+					array(
+						'status' => 'Not Reachable',
+						'message' => __('The Document page \'%s\' count not be reached.  It is possible that the Cacti Documentation site is not reachable.  The http error was \'%s\'.  Consider downloading an official release to obtain the latest page help documentation and hosting the documentation locally.', $page, $response[0])
+					)
+				);
+			} else {
+				print json_encode(
+					array(
+						'status' => 'Not Reachable',
+						'message' => __('The Document page \'%s\' count not be reached.  The Cacti Documentation site is not reachable.  Consider downloading an official release to obtain the latest page help documentation and hosting the documentation locally.', $page)
```
Comment (netniV) [inline help.php:81]: Wording is a little muddle, I feel (personally).  Consider:

> The Document page \'%s\' count not be reached on the Official Cacti Documentation website.

I feel this change removes the need for separate sentences (condensing) and more affirmation that we are trying to reach our own website.  But that's my personal opinion so may unnecessary.

File: help.php [hunk 5]
```diff
@@ -29,14 +29,76 @@
...
+				array(
+					'status'   => 'Success',
+					'location' => $config['url_path'] . '/docs/' . $page
+				)
+			);
+		} else {
+			print json_encode(
+				array(
+					'status' => 'Not Reachable',
+					'message' => __('The Document page \'%s\' count not be reached locally.  If you have installed from GitHub directly, consider downloading the distribution release and extracting the \'docs\' folder from that release to you Cacti install.', $page)
```
Comment (netniV) [inline help.php:98]: I feel like we should have a way to handle that.  For example, use of a parser for the files through the help page.  Also, so far, all of the connectivity attempted here is via the server, not the client so what's valid for the server may not be valid for the client and visa versa.

**Review by netniV — CHANGES_REQUESTED** [review]
The idea looks good in principle, but the more I think on it, the more it should really just be on the client-side.  

Consider this instead.  Creating a structure of possible URLs which is returned via the json call.  That list can vary dependant on the settings (like the one you've created).  These are then checked client side for validity in the order given.  If all fail, then we show a message when using multiple sources such as:

> The help page '%s' count not be reached on the Official Cacti Documentation website. or on the local server. Help can be found at %s.

Or alternatively, when the forced local setting is on:

> System settings prevent checking with official Cacti Group website, and Help page '%s' was not found locally. Help can be found at %s

Where the first parameter is the page and the second is the URL to it.  I am wondering if help pages should have their own array so we can use constants in the source.

**Comment by TheWitness** [comment]
Good points mark.  You know what's really embarrassing is that I never gave consideration that the actual page accessing was going to take place client side.  The assumption that if the server can not get the page, the user can not is kind of silly.  But people, including I make false assumptions all the time.  It's human nature.  Maybe the way to do is this.

1) Client knows the page, and the status of local only.
2) If local only is not set, the client attempts to access the page via callback.
3) Check the status code, if invalid, then do a second callback to either get the message, or the local page.

Outside of the wording changes, this does not seem too dramatic.

File: help.php (hunk 5 above)
Comment (TheWitness) [reply help.php:98]: Server argument agreed to.  How to automatically reconcile the "fetch it from XXX and install" I want to table for the moment.  If we host the HTML pages somewhere, we could for example, have a utility that pulls them separately, or even better, attempts to do so during the install process.  I would be okay deferring that for a subsequent change though.

File: help.php (hunk 3 above)
Comment (TheWitness) [reply help.php:61]: Yea, this is due to the documentation page never failing with a 404.  It always responds and then says the page does not exist.

File: help.php (hunk 2 above)
Comment (TheWitness) [reply help.php:54]: No, it always returns a 200 regardless if the page exists or not.

File: help.php (hunk 1 above)
Comment (TheWitness) [reply help.php:44]: Yea, I agree with that.  It was a raw copy and past for some of our other context options setups.

**Comment by TheWitness** [comment]
Well, this is lovely.  Can not access through a callback in that the server has a policy that forbids accessing it through a callback.

image

**Comment by TheWitness** [comment]
We should be able to relax that on the docs site don't you think?

**Comment by TheWitness** [comment]
Here is the revised message now that I fixed the URL.

image

**Comment by TheWitness** [comment]
The client side is going to be an issue with fall back as unless we change the security policy on docs.cacti.net to

```
Access-Control-Allow-Origin: *
```

Otherwise, there is no way to auto-detect the pages existence through a callback.  It does not seem such a radical change to me as it's essentially a public static site.

**Comment by TheWitness** [comment]
I went ahead and opened permissions a bit on the site.

File: help.php (hunk 5 above)
Comment (TheWitness) [reply help.php]: This is all being handled now.  I had to allow Ajax calls on the docs site, but now it's working.

File: help.php (hunk 4 above)
Comment (TheWitness) [reply help.php]: Message updated.  Still left directions, but message is now truncated.

File: help.php (hunk 1 above)
Comment (TheWitness) [reply help.php]: Done.

**Comment by TheWitness** [comment]
Everything updated now.  Please check the various error messages @Linegod

**Comment by TheWitness** [comment]
`Where the first parameter is the page and the second is the URL to it. I am wondering if help pages should have their own array so we can use constants in the source.`

We have an array, but it's in the get_help_page() function.  It might be interesting to make it an CONSTANT array, but I think that support starts in PHP7.

**Pull request merged**